managed/services/backup/removal_service.go

```diff
@@ -27,142 +31,300 @@ import (
 	"github.com/percona/pmm/managed/models"
 )
 
-// RemovalService manage removing of backup artifacts.
+// RemovalService manages removing of backup artifacts.
 type RemovalService struct {
-	l  *logrus.Entry
-	db *reform.DB
-	s3 s3
+	l              *logrus.Entry
+	db             *reform.DB
+	pbmPITRService pbmPITRService
 }
 
 // NewRemovalService creates new backup removal service.
-func NewRemovalService(db *reform.DB, s3 s3) *RemovalService {
+func NewRemovalService(db *reform.DB, pbmPITRService pbmPITRService) *RemovalService {
 	return &RemovalService{
-		l:  logrus.WithField("component", "services/backup/removal"),
-		db: db,
-		s3: s3,
+		l:              logrus.WithField("component", "services/backup/removal"),
+		db:             db,
+		pbmPITRService: pbmPITRService,
 	}
 }
 
-// DeleteArtifact deletes specified artifact.
-func (s *RemovalService) DeleteArtifact(ctx context.Context, artifactID string, removeFiles bool) error {
-	artifactName, s3Config, err := s.beginDeletingArtifact(artifactID)
+// DeleteArtifact deletes specified artifact along with files if specified.
+func (s *RemovalService) DeleteArtifact(storage Storage, artifactID string, removeFiles bool) error {
+	artifact, _, err := s.lockArtifact(artifactID, models.DeletingBackupStatus)
 	if err != nil {
 		return err
 	}
 
-	if s3Config != nil && removeFiles {
-		if err := s.s3.RemoveRecursive(
-			ctx,
-			s3Config.Endpoint,
-			s3Config.AccessKey,
-			s3Config.SecretKey,
-			s3Config.BucketName,
-			// Recursive listing finds all the objects with the specified prefix.
-			// There could be a problem e.g. when we have artifacts `backup-daily` and `backup-daily-1`, so
-			// listing by prefix `backup-daily` gives us both artifacts.
-			// To avoid such a situation we need to append a slash.
-			artifactName+"/"); err != nil {
-			if _, updateErr := models.UpdateArtifact(s.db.Querier, artifactID, models.UpdateArtifactParams{
-				Status: models.BackupStatusPointer(models.FailedToDeleteBackupStatus),
-			}); updateErr != nil {
-				s.l.WithError(updateErr).
-					Errorf("failed to set status %q for artifact %q", models.FailedToDeleteBackupStatus, artifactID)
-			}
+	defer func() {
+		if err != nil {
+			s.setArtifactStatus(artifactID, models.FailedToDeleteBackupStatus)
+		}
+	}()
```

Comment:
If you want to address `err` in defer than it should be declared in function. Otherwise if someone will add such code this func will be broken
```
if err = myFunc(); err != nil {
    l.Log(...)
}
return nil
```

Example: https://go.dev/play/p/rpfOO89WUyu

Reply:
Your example with `myFunc` only works if somebody also specify a name for returning value in func signature. Not sure this is a case, but as a code style good suggestion.